applications/IgaApplication/custom_modelers/refinement_modeler.cpp

```diff
@@ -0,0 +1,197 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License 
+//                   Kratos default license: kratos/license.txt
+//
+
+
+// Project includes
+#include "refinement_modeler.h"
+
+namespace Kratos
+{
+    ///@name Stages
+    ///@{
+
+    void RefinementModeler::PrepareGeometryModel()
+    {
+        const std::string DataFileName = mParameters.Has("refinements_file_name")
+            ? mParameters["refinements_file_name"].GetString()
+            : "refinements.iga.json";
+
+        KRATOS_INFO_IF("::[RefinementModeler]::", mEchoLevel > 0) << "Refining model by: " << DataFileName << std::endl;
+
+        const Parameters refinements_parameters = ReadParamatersFile(DataFileName);
+
+        ApplyRefinements(refinements_parameters);
+    }
+
+    void RefinementModeler::ApplyRefinements(
+        const Parameters rParameters) const
+    {
+        KRATOS_ERROR_IF_NOT(rParameters.Has("refinements"))
+            << "Parameters do not have refinements section.\n"
+            << rParameters << std::endl;
+
+        KRATOS_ERROR_IF_NOT(rParameters["refinements"].IsArray())
+            << "refinements section need to be of type array.\n"
+            << rParameters << std::endl;
+
+        for (IndexType i = 0; i < rParameters["refinements"].size(); ++i) {
+            ApplyRefinement(rParameters["refinements"][i]);
+        }
+    }
+
+    void RefinementModeler::ApplyRefinement(
+        const Parameters rParameters) const
+    {
+        KRATOS_ERROR_IF_NOT(rParameters.Has("model_part_name"))
+            << "Missing \"model_part_name\" in refinements block.\n"
+            << rParameters << std::endl;
+        const std::string model_part_name = rParameters["model_part_name"].GetString();
+        ModelPart& r_model_part = mpModel->HasModelPart(model_part_name)
+            ? mpModel->GetModelPart(model_part_name)
+            : mpModel->CreateModelPart(model_part_name);
+
+        // Generate the list of geometries, which are needed, here.
+        GeometriesArrayType geometry_list;
+        GetGeometryList(geometry_list, r_model_part, rParameters);
+
+        KRATOS_ERROR_IF_NOT(rParameters.Has("parameters"))
+            << "Missing \"parameters\" in refinements block.\n"
+            << rParameters << std::endl;
+        KRATOS_ERROR_IF_NOT(rParameters.Has("geometry_type"))
+            << "Missing \"geometry_type\".\n"
+            << rParameters << std::endl;
+        if (rParameters["geometry_type"].GetString() == "NurbsSurface") {
```

Comment:
Do we wanna enter a new function here? Then I will also adjust the volume refinement to be called from here.

Reply:
would make sense ;)